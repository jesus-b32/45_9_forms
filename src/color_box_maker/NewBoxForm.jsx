import React, { useState } from "react";
import { v4 as uuid } from 'uuid';

/** Form for creating a new box to add to a list.
 *
 * Has state for the backgroundColor, width, and height of the box; 
 * on submission, sends {backgroundColor, width, height} to fn rec'd from parent.
 *
 */

function NewBoxForm ({ addBox }) {
    const INITIAL_STATE = {
        backgroundColor: '',
        width: '',
        height: ''
    }

    /**
     * formDate is an object that holds 
     *  backgroundColor
     *  width
     *  height
     *  */
    const [formData, setFormData] = useState(INITIAL_STATE);

    /** Update local state w/curr state of input elem */
    function handleChange (event) {
        const { name, value } = event.target;
        setFormData(formData => ({
            ...formData,
            [name]: value
        }))
    }

    /** Send {backgroundColor, width, height} to parent
    *    & clear form. */
    function handleSubmit (event) {
        event.preventDefault();
        addBox({ ...formData, id: uuid() });
        setFormData(INITIAL_STATE);
    }

    /** render form */
    return (
        <form onSubmit={handleSubmit}>
            <label htmlFor="backgroundColor">Background Color: </label>
            <input
                id="backgroundColor"
                type="text"
                name="backgroundColor"
                placeholder="Background Color"
                value={formData.backgroundColor}
                onChange={handleChange}
            />
            <label htmlFor="width">Width: </label>
            <input
                id="width"
                type="number"
                min={1}
                max={25}
                name="width"
                placeholder="Width"
                value={formData.width}
                onChange={handleChange}
            />
            <label htmlFor="height">Height: </label>
            <input
                id="height"
                type="number"
                min={1}
                max={25}
                name="height"
                placeholder="Height"
                value={formData.height}
                onChange={handleChange}
            />
            <button>Add a new Box</button>
        </form>
    )

}

export default NewBoxForm;