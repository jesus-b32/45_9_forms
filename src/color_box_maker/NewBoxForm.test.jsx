import React from "react";
import { render, fireEvent } from "@testing-library/react";
import NewBoxForm from "./NewBoxForm";

describe('NewBoxForm', () => {
    //smoke test
    it('renders App component', () => {
        render(<NewBoxForm/>);
    })

    //snapshot test
    it('matches snapshot', () => {
        const {asFragment} = render(<NewBoxForm/>);
        expect(asFragment()).toMatchSnapshot();
    })

    it("runs the addBox function on form submit", function() {
        const createMock = vi.fn();
        const { getByText } = render(<NewBoxForm addBox={createMock} />);
        const createButton = getByText("Add a new Box");
        fireEvent.click(createButton);
        expect(createMock).toHaveBeenCalled();
    });
})
