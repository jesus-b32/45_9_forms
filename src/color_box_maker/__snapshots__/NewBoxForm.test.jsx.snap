// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NewBoxForm > matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <label
      for="backgroundColor"
    >
      Background Color: 
    </label>
    <input
      id="backgroundColor"
      name="backgroundColor"
      placeholder="Background Color"
      type="text"
      value=""
    />
    <label
      for="width"
    >
      Width: 
    </label>
    <input
      id="width"
      max="25"
      min="1"
      name="width"
      placeholder="Width"
      type="number"
      value=""
    />
    <label
      for="height"
    >
      Height: 
    </label>
    <input
      id="height"
      max="25"
      min="1"
      name="height"
      placeholder="Height"
      type="number"
      value=""
    />
    <button>
      Add a new Box
    </button>
  </form>
</DocumentFragment>
`;
